scanner/config.go

```diff
@@ -50,94 +80,154 @@ func Config(searchDir string) models.ScanResultModel {
 
 	//
 	// Scan
-	projectScanners := scanners.ActiveScanners
+	log.TInfof(colorstring.Blue("Running scanners:"))
+	fmt.Println()
 
-	projectTypeErrorMap := map[string]models.Errors{}
-	projectTypeWarningMap := map[string]models.Warnings{}
-	projectTypeOptionMap := map[string]models.OptionModel{}
-	projectTypeConfigMap := map[string]models.BitriseConfigMap{}
+	// Collect scanner outputs, by scanner name
+	scannerToOutput := map[string]scannerOutput{}
+	{
+		projectScannerMatchResults := mapScannersToOutput(scanners.ProjectScanners, searchDir)
+		detectedProjectTypes := getDetectedScannerNames(projectScannerMatchResults)
+		log.Printf("Detected project types: %s", detectedProjectTypes)
+		fmt.Println()
 
-	excludedScannerNames := []string{}
+		// Project types are needed by tool scanners, to create decision tree on which project type
+		// to actually use in bitrise.yml
+		if len(detectedProjectTypes) == 0 {
+			detectedProjectTypes = []string{otherProjectType}
+		}
+		for _, toolScanner := range scanners.AutomationToolScanners {
+			toolScanner.(scanners.AutomationToolScanner).SetDetectedProjectTypes(detectedProjectTypes)
+		}
 
-	log.TInfof(colorstring.Blue("Running scanners:"))
-	fmt.Println()
+		toolScannerResults := mapScannersToOutput(scanners.AutomationToolScanners, searchDir)
+		detectedAutomationToolScanners := getDetectedScannerNames(toolScannerResults)
+		log.Printf("Detected automation tools: %s", detectedAutomationToolScanners)
+		fmt.Println()
 
-	for _, detector := range projectScanners {
-		detectorName := detector.Name()
-		detectorWarnings := []string{}
-		detectorErrors := []string{}
+		// Merge project and tool scanner outputs
+		scannerToOutput = toolScannerResults
+		for k, v := range projectScannerMatchResults {
+			scannerToOutput[k] = v
+		}
+	}
 
-		log.TInfof("Scanner: %s", colorstring.Blue(detectorName))
+	scannerToWarnings := map[string]models.Warnings{}
+	scannerToErrors := map[string]models.Errors{}
+	scannerToOptions := map[string]models.OptionNode{}
+	scannerToConfigMap := map[string]models.BitriseConfigMap{}
+	for k, v := range scannerToOutput {
+		if v.scanResult == scanResultNotDetected && v.warnings != nil ||
```

Comment:
why do you check for nil here and why not in the following case?

Reply:
added comment